hack/make-rules/test.sh

```diff
@@ -180,113 +173,72 @@ junitFilenamePrefix() {
-  if [[ -z "${junit_filename_prefix}" ]]; then
-    return
-  fi
-
-  local junit_xml_filename
-  junit_xml_filename="${junit_filename_prefix}.xml"
-
+installTools() {
   if ! command -v gotestsum >/dev/null 2>&1; then
     kube::log::status "gotestsum not found; installing from ./hack/tools"
     go -C "${KUBE_ROOT}/hack/tools" install gotest.tools/gotestsum
   fi
-  gotestsum --junitfile "${junit_xml_filename}" --raw-command cat "${junit_filename_prefix}"*.stdout
-  if [[ ! ${KUBE_KEEP_VERBOSE_TEST_OUTPUT} =~ ^[yY]$ ]]; then
-    rm "${junit_filename_prefix}"*.stdout
-  fi
 
   if ! command -v prune-junit-xml >/dev/null 2>&1; then
     kube::log::status "prune-junit-xml not found; installing from ./cmd"
     go -C "${KUBE_ROOT}/cmd/prune-junit-xml" install .
   fi
-  prune-junit-xml "${junit_xml_filename}"
-
-  kube::log::status "Saved JUnit XML test report to ${junit_xml_filename}"
 }
 
 runTests() {
   local junit_filename_prefix
   junit_filename_prefix=$(junitFilenamePrefix)
 
-  # Try to normalize input names.
+  installTools
+
+  # Try to normalize input names. This is slow!
   local -a targets
+  kube::log::status "Normalizing Go targets"
   kube::util::read-array targets < <(kube::golang::normalize_go_targets "$@")
 
-  # If we're not collecting coverage, run all requested tests with one 'go test'
-  # command, which is much faster.
-  if [[ ! ${KUBE_COVER} =~ ^[yY]$ ]]; then
-    kube::log::status "Running tests without code coverage ${KUBE_RACE:+"and with ${KUBE_RACE}"}"
-    # shellcheck disable=SC2031
-    go test "${goflags[@]:+${goflags[@]}}" \
-     "${KUBE_TIMEOUT}" "${targets[@]}" \
-     "${testargs[@]:+${testargs[@]}}" \
-     | tee ${junit_filename_prefix:+"${junit_filename_prefix}.stdout"} \
-     | grep --binary-files=text "${go_test_grep_pattern}" && rc=$? || rc=$?
-    produceJUnitXMLReport "${junit_filename_prefix}"
-    return "${rc}"
+  # Enable coverage data collection?
+  local cover_msg
+  local COMBINED_COVER_PROFILE
+
+  if [[ ${KUBE_COVER} =~ ^[yY]$ ]]; then
+    cover_msg="with code coverage"
+    if [[ -z "${KUBE_COVER_REPORT_DIR}" ]]; then
+      cover_report_dir="/tmp/k8s_coverage/$(kube::util::sortable_date)"
+    else
+      cover_report_dir="${KUBE_COVER_REPORT_DIR}"
+    fi
+    kube::log::status "Saving coverage output in '${cover_report_dir}'"
+    mkdir -p "${@+${@/#/${cover_report_dir}/}}"
+    COMBINED_COVER_PROFILE="${cover_report_dir}/combined-coverage.out"
+    goflags+=(-cover -covermode="${KUBE_COVERMODE}" -coverprofile="${COMBINED_COVER_PROFILE}")
+  else
+    cover_msg="without code coverage"
   fi
 
-  kube::log::status "Running tests with code coverage ${KUBE_RACE:+"and with ${KUBE_RACE}"}"
+  kube::log::status "Running tests ${cover_msg} ${KUBE_RACE:+"and with ${KUBE_RACE}"}"
+  gotestsum --format="${gotestsum_format}" \
+            --jsonfile="$(if [[ -n "${junit_filename_prefix}" ]] && [[ ${KUBE_KEEP_VERBOSE_TEST_OUTPUT} =~ ^[yY]$ ]]; then echo "${junit_filename_prefix}.stdout"; fi)" \
```

Comment:
this inlining is 🤯

Reply:
I can move it into a variable if you prefer that.

Reply:
@BenTheElder and @thockin are the bash reviewers, I personally prefer to move to a variable because  to me is simpler to read and  is harder to have syntax mistakes caused by bash expansions or similar, there are a lot of symbols, quotes and things here

Reply:
Done.